### PR TITLE
 Add 'inherit' and 'initial' CSS properties

### PR DESCRIPTION
**Pull request by embeddedt**
### Description of the feature or fix

Implements the style values suggested in #3086.

Concerns:

* The API surface has tripled due to the need for separate APIs to set `inherit` and `initial`. It may be better to go with @FoundationKen's original suggestion and require users to use `lv_style_set_prop_meta` directly. Since there aren't three possible value types, there is no type safety issue there.
* `initial` currently uses the default value of a property, however, it might be more intuitive to derive that from the currently selected theme instead.
* No documentation yet.

### Checkpoints
- [ ] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Run `code-format.py` from the `scripts` folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the documentation

### CODE REVIEW

File: src/core/lv_obj_style.c [hunk 1]
```diff
@@ -224,14 +224,14 @@ lv_style_value_t lv_obj_get_style_prop(const lv_obj_t * obj, lv_part_t part, lv_
 {
     lv_style_value_t value_act;
     bool inherit = lv_style_prop_has_flag(prop, LV_STYLE_PROP_INHERIT);
```
Comment (kisvegabor) [inline src/core/lv_obj_style.c]: Do we need to store the  `INHERIT` flag for each property or should it be rather the default value of some properties?

File: src/misc/lv_style.c [hunk 2]
```diff
@@ -155,7 +162,7 @@ void lv_style_reset(lv_style_t * style)
 {
     LV_ASSERT_STYLE(style);
 
-    if(style->is_const) {
+    if(style->prop1 == LV_STYLE_PROP_ANY) {
```
Comment (kisvegabor) [inline src/misc/lv_style.c:168]: What is the purpose of this change?

**Review by kisvegabor — COMMENTED** [review]
> The API surface has tripled due to the need for separate APIs to set inherit and initial. It may be better to go with @FoundationKen's original suggestion and require users to use lv_style_set_prop_meta directly. Since there aren't three possible value types, there is no type safety issue there.

I agree to use a common function to set meta things. as these settings probably will be only rarely used. It makes the API a little bit inconsistent, but I think it's the best we can do in C. 

> initial currently uses the default value of a property, however, it might be more intuitive to derive that from the currently selected theme instead.

Great idea! I was already to thinking about how to handle the themes in a more clever way. My idea is that there could be 3 level of styles:
- local style: we already know it
- normal styles (like css classes): we also know it
- themes: unlike now, it can be a totally separated layer. If a property wasn't found in local or normal styles LVGL says: "okay theme, your turn, what should it be if no one knows." (i.e. it's the default).

This way a theme is not a huge collection of styles which are added to all objects by default, but just callback like:
```
lv_style_prop_t theme_get_prop(obj, prop, part, state);
``` 

Under the hood it can has a large switch case to find a style from where it gets the property, but it can also just return a color, width, radius, etc.

Moreover, it can return context based values too. E.g. for every 2nd button return gray, and white for the others. Or if the parent of a slider is a tabview, use different values. Basically what CSS selectors can do. It can be like the browser's style sheet. 
We can also let the users to specify a callback which will be called before the theme's callback to easily hook the default theme.

This lead a little bit far, but this way we could handle the styles a more optimal and flexible way. And getting the default value would be obvious too.

File: src/core/lv_obj_style.c (hunk 1 above)
Comment (embeddedt) [reply src/core/lv_obj_style.c]: We could have a hardcoded default value but I suspect the added code will use up most of the space saved by not having a lookup table for built-in properties.

File: src/misc/lv_style.c (hunk 2 above)
Comment (embeddedt) [reply src/misc/lv_style.c:168]: `is_const` had to be removed so that the address of `prop1` can be taken. Before, it was a bitfield. Since `LV_STYLE_PROP_ANY` isn't a valid value to store in `prop1`, I used that as a check for a constant style instead.

File: src/core/lv_obj_style.c (hunk 1 above)
Comment (kisvegabor) [reply src/core/lv_obj_style.c]: I rather concerned about the inconsistency as this way we can set the inherit flag in 2 places:
- in the [prop look up table](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_style.c#L29) 
- in the property value (with this feature)

Once we have this feature we don't need to use the look up table for this as it causes confusion about where inheritance is/should be set. 

Adding ~10 props [here](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_style.c#L361) shouldn't require too much space but at least there were only one clear way to set inherit flag.

File: src/misc/lv_style.c (hunk 2 above)
Comment (kisvegabor) [reply src/misc/lv_style.c:168]: In this case can we add dedicated name for const style? E.g. `_LV_STYLE_PROP_CONST`.

File: src/core/lv_obj_style.c (hunk 1 above)
Comment (embeddedt) [reply src/core/lv_obj_style.c]: These are separate flags. The flag here (`LV_STYLE_PROP_INHERIT`, which will be renamed to `LV_STYLE_PROP_FLAG_INHERIT` in v9) specifies that the property supports inheritance from parents at all (because some like x and y don't). That flag is specified at the property type level.

The other flag (the new one) is effectively a special value for each style that forces `get_prop_core` to immediately check the parent's styles. That is specified in the prop-value pair within a style. It just so happens to be stored with the property ID because there are free bits there.

Comment (kisvegabor) [reply src/core/lv_obj_style.c]: >  The flag here (LV_STYLE_PROP_INHERIT, which will be renamed to LV_STYLE_PROP_FLAG_INHERIT in v9) specifies that the property supports inheritance from parents at all (because some like x and y don't). 

I see. Then what about a name like `LV_STYLE_PROP_FLAG_INHERITABLE`?

Comment (embeddedt) [reply src/core/lv_obj_style.c]: That name is definitely better, but we can't change it till v9 without breaking user code (unless we keep the old name as well).

Or do you think it's minor enough to change, since it can be fixed with a find-and-replace?

Comment (kisvegabor) [reply src/core/lv_obj_style.c]: We can keep it, I just wanted to be sure we are talking about the same thing. I added a note about the new name here: https://github.com/lvgl/lvgl/pull/3390#discussion_r885915769 

I suggest changing at least the variable name:
```suggestion
    bool inheritable = lv_style_prop_has_flag(prop, LV_STYLE_PROP_INHERIT);
```

File: src/misc/lv_style.c (hunk 2 above)
Comment (embeddedt) [reply src/misc/lv_style.c:168]: I've added it, and under the hood it uses the same value for now.

File: src/core/lv_obj_style.c (hunk 1 above)
Comment (embeddedt) [reply src/core/lv_obj_style.c]: Changed as you suggested.

**Comment by kisvegabor** [comment]
Looks good! The only missing thing is to write a few sentences about it in the docs.
